File: site/__tests__/tbdex/pfi/pfi-overview.snippet.kt
// :snippet-start: pfiOverviewConfigKt
import tbdex.sdk.httpserver.TbdexHttpServer
import web5.sdk.crypto.AwsKeyManager
import web5.sdk.dids.methods.dht.DidDht
import my.project.ExchangesApiProvider
import my.project.OfferingsApiProvider 

val serverConfig = TbdexHttpServerConfig(
    port = 8080,
    pfiDid = DidDht.create(AwsKeyManager()).uri,
    offeringsApi = ExchangesApiProvider(),
    exchangesApi = OfferingsApiProvider()
  )

val httpApi = TbdexHttpServer(serverConfig)
// :snippet-end:

// :snippet-start: pfiOverviewServerRoutesKt
import tbdex.sdk.httpserver.models.SubmitKind

httpApi.submit(SubmitKind.rfq) { call, messageKind, offering ->
    ExchangesApiProvider.write(offering)
    call.respond(HttpStatusCode.Accepted)
}

httpApi.submit(SubmitKind.order) { call, messageKind, offering ->
    ExchangesApiProvider.write(offering)
    call.respond(HttpStatusCode.Accepted)
}

httpApi.submit(SubmitKind.close) { call, messageKind, offering ->
    ExchangesApiProvider.write(offering)
    call.respond(HttpStatusCode.Accepted)
}
// :snippet-end:

// :snippet-start: pfiOverviewServerStartKt
httpApi.start()
// :snippet-end:

// :snippet-start: pfiOverviewWriteKt

// :snippet-end:

// :snippet-start: pfiOverviewReadOfferingsKt

// :snippet-end:

// :snippet-start: pfiOverviewWriteOfferingsKt

// :snippet-end:

File: site/__tests__/tbdex/pfi/pfi-overview.snippet.js
// :snippet-start: pfiOverviewConfigJs
import { TbdexHttpServer } from '@tbdex/http-server'

const httpApi = new TbdexHttpServer({ exchangesApi: ExchangesApiProvider, offeringsApi: OfferingsApiProvider })
// :snippet-end:

// :snippet-start: pfiOverviewServerRoutesJs
import {
    Rfq,
    Order,
    Close
} from '@tbdex/http-server'

httpApi.submit('rfq', async (ctx, rfq) => {
    await ExchangesApiProvider.write({ message: rfq})
})

httpApi.submit('order', async (ctx, order) => {
    await ExchangesApiProvider.write({ message: order })
})

httpApi.submit('close', async (ctx, close) => {
    await ExchangesApiProvider.write({ message: close })
})
// :snippet-end:

// :snippet-start: pfiOverviewServerStartJs
const server = httpApi.listen(config.port, () => {
    log.info(`Mock PFI listening on port ${config.port}`)
})
// :snippet-end:

// :snippet-start: pfiOverviewWriteJs
async function write({ message }) {
    const result = await dataProvider.insert(
        'exchange',
        {
            exchangeid: message.exchangeId,
            messagekind: message.kind,
            messageid: message.id,
            subject,
            message: JSON.stringify(message)
        });
}  
// :snippet-end:

// :snippet-start: pfiOverviewReadOfferingsJs
import { Offering } from '@tbdex/http-server'

async function getOffering(opts) {
  const [ result ] =  await dataProvider.queryForOffering('offering', opts.id);

  if (!result) {
    return undefined
  }
  return Offering.factory(result.offering)
}


async function getOfferings() {
  const results =  await dataProvider.getOfferings('offering');
  const offerings = []

  for (let result of results) {
    const offering = Offering.factory(result.offering)
    offerings.push(offering)
  }

  return offerings
}
// :snippet-end:

// :snippet-start: pfiOverviewWriteOfferingsJs
async function create(offering) {
    let result = await dataProvider.insert(
        'offering',
        {
        offeringid: offering.id,
        payoutcurrency: offering.payoutCurrency.currencyCode,
        payincurrency: offering.payinCurrency.currencyCode,
        offering: JSON.stringify(offering)
    });
}
// :snippet-end:
